teps/0037-simplify-metrics.md

```diff
@@ -0,0 +1,226 @@
+
+### Use Cases
+
+Cluster admin users are able to configure tekton pipelines to produce high level metrics and focus on overall status to make sure Tekton works fine. While single users who care about statistics of indivial taskrun can still keep the fine grained metrics.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Coarse grained metrics should be provided. This can be implemented in 2 ways: changing taskRun level metrics to namespace level and changing metrics type. These 2 methods is mutally exclusive.
+If users care about individual taskRun performance, they can set `duration seconds` type from `histogram` to `gauge`. If users care about overall performance, they can collect metrics at namespace level.
+
+### Change taskRun level metrics to namespace level
+
+We can add a `config-observability` option to switch between taskRun level metrics and namespace level.
+`metrics.namespace-level` field indicates whether to aggregate metrics at namespace level. When set to `true`, it will remove taskRun label in metrics. Sizable reduction in metrics count can happen when multiple taskRuns is running in the same namespace.
+By default, it is set to `false`.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-observability
+  namespace: tekton-pipelines
+  labels:
+    app.kubernetes.io/instance: default
+    app.kubernetes.io/part-of: tekton-pipelines
+data:
+  metrics.namespace-level: "true"
+```
+
+When the option is `false`, every taskrun in a namespace has a seperate metrics to record its running status.
+
+```log
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="anonymous",taskrun="duplicate-pod-task-run-ytqrdxja",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="anonymous",taskrun="duplicate-pod-task-run-ymelobwl",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="anonymous",taskrun="duplicate-pod-task-run-xnuasulj",le="10"} 1
+```
+
+When the option is `true`, these metrics will be merged into one.
+
+```log
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",le="10"} 3
+```
+
+### Change metrics type
+
+Large metrics volumn is mostly caused by histogram metrics. Single histogram metrics will produce 15 metrics. These metrics type could be changed to gauge which would reduce 15 metrics to one.
+For example, metrics like `tekton_pipelinerun_duration_seconds`, `taskrun_duration_seconds`, `tekton_pipelinerun_taskrun_duration_seconds` are histogram. But the metrics could not provide much information at taskrun level and can be changed to `gauge`.
```

Comment:
Additionally `tekton_pipelinerun_taskrun_duration_seconds` could be dropped. `taskrun_duration_seconds` is sufficient.

Reply:
`tekton_pipelinerun_taskrun_duration_seconds` is used to identify if a taskrun is an individual taskrun or from a pipelinerun. This metrics and `tekton_taskrun_duration_seconds` are mutually exclusive. So there won't be any decrease in metrics amount if we drop `tekton_pipelinerun_taskrun_duration_seconds`.